Firestore/Source/API/FIRDocumentChange.mm

```diff
@@ -14,128 +14,24 @@
-      HARD_ASSERT(change.type() == DocumentViewChange::Type::kAdded,
-                  "Invalid event type for first snapshot");
-      HARD_ASSERT(!lastDocument || snapshot.query().comparator(lastDocument, change.document()) ==
-                                       NSOrderedAscending,
-                  "Got added events in wrong order");
-      [changes addObject:[[FIRDocumentChange alloc] initWithType:FIRDocumentChangeTypeAdded
-                                                        document:document
-                                                        oldIndex:NSNotFound
-                                                        newIndex:index++]];
-    }
-    return changes;
-  } else {
-    // A DocumentSet that is updated incrementally as changes are applied to use to lookup the index
-    // of a document.
-    DocumentSet indexTracker = snapshot.old_documents();
-    NSMutableArray<FIRDocumentChange *> *changes = [NSMutableArray array];
-    for (const DocumentViewChange &change : snapshot.document_changes()) {
-      if (!includeMetadataChanges && change.type() == DocumentViewChange::Type::kMetadata) {
-        continue;
-      }
-
-      FIRQueryDocumentSnapshot *document = [[FIRQueryDocumentSnapshot alloc]
-          initWithFirestore:firestore
-                documentKey:change.document().key
-                   document:change.document()
-                  fromCache:snapshot.from_cache()
-           hasPendingWrites:snapshot.mutated_keys().contains(change.document().key)];
-
-      size_t oldIndex = DocumentSet::npos;
-      size_t newIndex = DocumentSet::npos;
-      if (change.type() != DocumentViewChange::Type::kAdded) {
-        oldIndex = indexTracker.IndexOf(change.document().key);
-        HARD_ASSERT(oldIndex != DocumentSet::npos, "Index for document not found");
-        indexTracker = indexTracker.erase(change.document().key);
-      }
-      if (change.type() != DocumentViewChange::Type::kRemoved) {
-        indexTracker = indexTracker.insert(change.document());
-        newIndex = indexTracker.IndexOf(change.document().key);
-      }
-      [FIRDocumentChange documentChangeTypeForChange:change];
-      FIRDocumentChangeType type = [FIRDocumentChange documentChangeTypeForChange:change];
-      [changes addObject:[[FIRDocumentChange alloc] initWithType:type
-                                                        document:document
-                                                        oldIndex:oldIndex
-                                                        newIndex:newIndex]];
-    }
-    return changes;
-  }
-}
-
-@end
-
-@implementation FIRDocumentChange
-
-- (instancetype)initWithType:(FIRDocumentChangeType)type
-                    document:(FIRQueryDocumentSnapshot *)document
-                    oldIndex:(NSUInteger)oldIndex
-                    newIndex:(NSUInteger)newIndex {
+- (instancetype)initWithDocumentChange:(DocumentChange)documentChange {
```

Comment:
Optional: I think you could take `DocumentChange` by rvalue reference here: `DocumentChange&&`. That way, the compiler would enforce that a `DocumentChange` must be moved into this constructor. I don't think there's a valid case for creating a `FIRDocumentChange` from a `DocumentChange` without transferring ownership, but feel free to correct me.

Reply:
`SnapshotMetadata` probably doesn't use `std::move` because moving is not more efficient than copying for the class (it consists of primitives).

The way I see it is that this would be an unreasonable restriction (and very annoying) if there was a valid case for passing a copy. If the ownership is always transferred, I would say making the parameter an rvalue helps to use the API correctly by preventing a class of silent performance errors (ok, it doesn't matter here, but it may matter in the general case).

I I were to try to define a guideline, I'd say a parameter should be an rvalue reference if the ownership is always transferred and it's desirable to make the compiler prevent unintentional copies. A valid counterargument could be made that this introduces additional complexity that is not outweighed by the performance gain. Note that the style guide in general wants to minimize the use of rvalue references and would prefer passing by value (see [go/totw/117](go/totw/117)). To summarize, that's the approach I'm leaning towards, but I don't feel strongly about it, and I'm fine if you disagree and revert to the previous state.

Reply:
Thanks. I've gone ahead and removed it as it does seem unnecessary and go/cpp-primer specifically says you shouldn't use it "just in case", but I'm also starting to think we might overuse it in other places.  🤷‍♂️